File: exo/cpp/src/haptics.cpp
#include "exo_example_support.hpp"

#include <chrono>
#include <cstdlib>
#include <exception>
#include <iostream>
#include <string>
#include <thread>

int main(int argc, char** argv) {
    try {
        std::string port;
        std::string name = "primary";
        std::uint8_t left = 60;
        std::uint8_t right = 60;
        double hold = 1.0;
        for (int index = 1; index < argc; ++index) {
            const std::string argument = argv[index];
            if (argument == "--port") {
                port = example::require_value(argc, argv, index, argument);
            } else if (argument == "--name") {
                name = example::require_value(argc, argv, index, argument);
            } else if (argument == "--left") {
                left = example::parse_strength(
                    example::require_value(argc, argv, index, argument),
                    argument);
            } else if (argument == "--right") {
                right = example::parse_strength(
                    example::require_value(argc, argv, index, argument),
                    argument);
            } else if (argument == "--hold") {
                hold = example::parse_seconds(
                    example::require_value(argc, argv, index, argument));
            } else {
                throw std::invalid_argument("unknown argument: " + argument);
            }
        }
        if (port.empty()) throw std::invalid_argument("--port is required");

        qnbot::Sdk sdk(example::serial_config(port, name));
        auto exo = sdk.exo();
        auto device = exo.device(name);
        if (!device.get_device_info().capabilities.haptics) {
            std::cout << "this device reports no haptics" << std::endl;
            exo.close();
            return EXIT_SUCCESS;
        }

        exo.start();
        auto haptics = device.haptics();
        qnbot::ExoHaptics value;
        value.left = left;
        value.right = right;
        haptics.set(value);
        std::cout << "set left=" << static_cast<int>(left)
                  << " right=" << static_cast<int>(right) << std::endl;
        std::this_thread::sleep_for(std::chrono::duration<double>(hold));

        haptics.clear();
        std::cout << "haptics cleared" << std::endl;
        exo.stop();
        exo.close();
        return EXIT_SUCCESS;
    } catch (const std::exception& error) {
        return example::report_error(error);
    }
}

File: exo/python/src/haptics.py
from __future__ import annotations

import argparse
import time

from qnbot_sdk import Sdk, SerialConnection
from qnbot_sdk.exo import ExoConfig, ExoHaptics


def create_sdk(port: str, name: str) -> Sdk:
    return Sdk(devices=(ExoConfig(name=name, connection=SerialConnection(port=port)),))


def main() -> None:
    arguments = argparse.ArgumentParser(description="Set and clear Exo haptics")
    arguments.add_argument("--port", required=True, help="Serial port for the device")
    arguments.add_argument("--name", default="primary", help="Logical device name")
    arguments.add_argument("--left", type=int, default=60)
    arguments.add_argument("--right", type=int, default=60)
    arguments.add_argument("--hold", type=float, default=1.0)
    options = arguments.parse_args()

    sdk = create_sdk(options.port, options.name)
    exo = sdk.exo()
    device = exo.device(options.name)
    if not device.get_device_info().capabilities.haptics:
        print("this device reports no haptics", flush=True)
        exo.close()
        return

    exo.start()
    haptics = device.haptics()
    haptics.set(ExoHaptics(left=options.left, right=options.right))
    print(
        f"set left={options.left} right={options.right}",
        flush=True,
    )
    time.sleep(options.hold)

    haptics.clear()
    print("haptics cleared", flush=True)
    exo.stop()
    exo.close()


if __name__ == "__main__":
    main()
